# Project 1

## 1

A package is a group of Python modules that, when imported into your Python workspace allow for the use of various, helpful commands. Within each package are a series of libraries that each have a specific function relating to the main package. To install a package and library, you must first make sure the intended package is installed within the Python interpreter - found under Preferences in PyCharm. If it is not installed, a package can be easily found by typing in its name and adding it to the interpreter. Once the package is installed, you then have to import it into the workspace with a line of code. 

In [3]:
import pandas as pd

This would import the pandas library into your workspace under an alias 'pd'. Using an alias can be helpful in keeping code short and easy to read.

In [4]:
from datetime import datetime

This would import the datetime function from the datetime library. This is helpful when you only need access to a certain function from within a library; rather than importing the entire library, you can import certain functions specifically. 

## 2

A data frame is a structure that allows for the reading of data sets by organizing data entries into a table of columns and values. When working with dataframes, the pandas library is particularly helpful as it allows for the reading of the data in addition to data subsetting and manipulation. To read a file in its remote location in your file system, you would need to write a `read_()` command, specifying the library you are using and the path to the file. The pandas library has already been imported under an alias, so the code used would look something like this:

In [5]:
path_to_data = 'gapminder.tsv'

In [6]:
data = pd.read_csv(path_to_data, sep = '\t')

The dataset we're working this is tab-separated; however the pandas library will assume it is a comma-separated file. To get around this, within the `read_()` command, specify that you want data read as a csv file, `read_csv()`, while including that the data originally was tab-separated, `sep = '\t'`. 

To return a description of the data, use the `describe()` command:

In [22]:
data.describe()

year      lifeExp           pop      gdpPercap
count  1704.00000  1704.000000  1.704000e+03    1704.000000
mean   1979.50000    59.474439  2.960121e+07    7215.327081
std      17.26533    12.917107  1.061579e+08    9857.454543
min    1952.00000    23.599000  6.001100e+04     241.165877
25%    1965.75000    48.198000  2.793664e+06    1202.060309
50%    1979.50000    60.712500  7.023596e+06    3531.846989
75%    1993.25000    70.845500  1.958522e+07    9325.462346
max    2007.00000    82.603000  1.318683e+09  113523.132900

To determine how many rows and columns are included in the dataset, use the .shape() command.

In [28]:
data.shape[0]

1704

In [29]:
data.shape[1]

6

There are 1704 rows and 6 columns in this dataframe. To get a summary of the names of the columns, use `data.info()`, or `data.columns()` as a list:

In [30]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1704 entries, 0 to 1703
Data columns (total 6 columns):
country      1704 non-null object
continent    1704 non-null object
year         1704 non-null int64
lifeExp      1704 non-null float64
pop          1704 non-null int64
gdpPercap    1704 non-null float64
dtypes: float64(2), int64(2), object(2)
memory usage: 80.0+ KB


In [31]:
list(data.columns)

['country', 'continent', 'year', 'lifeExp', 'pop', 'gdpPercap']

## 3

In [32]:
data['year'].unique().tolist()

[1952, 1957, 1962, 1967, 1972, 1977, 1982, 1987, 1992, 1997, 2002, 2007]

In [33]:
len(data[(data['year'] == data['year'].max())])

142

In [34]:
len(data[(data['year'] == 2002)])

142

Starting in 1952, the years come in five-year increments. To make it more current, data from 2012 and 2017 should be included, adding 142 records for each year to the data frame - 284 in total.

## 4

In [35]:
data[(data['lifeExp'] == data['lifeExp'].min())]

country continent  year  lifeExp      pop   gdpPercap
1292  Rwanda    Africa  1992   23.599  7290203  737.068595


In this data frame, the lowest life expectancy occured in Rwanda in 1992, where life expectancy was about 23 years. This was around the time of the Rwanda genocide, when up to a million people were killed. This, in addition to the nation's poor public health at the time was likely the biggest factor in driving down average life expectancy. 

## 5

In [ ]:
total_gdp = (data['pop'] * data['gdpPercap'])
data['totalGDP'] = total_gdp.tolist()


In 2007, total GDP for Germany, France, Italy, and Spain were as follows: 

In [ ]:
data_europe2007 = data[(data['continent']=='Europe') & (data['year'] == data['year'].max())]
data_fgis2007 = data_europe2007[(data_europe2007['country']=='Spain') | (data_europe2007['country']=='France') | (data_europe2007['country']=='Germany') | (data_europe2007['country']=='Italy')]
name_gdp = data_fgis2007[['country', 'totalGDP']]
name_gdp.sort_values(by=['totalGDP'], ascending = False)

In 2002, total GDP for Germany, France, Italy, and Spain were as follows: 

In [ ]:
data_europe2002 = data[(data['continent']=='Europe') & (data['year'] == 2002)]
data_fgis2002 = data_europe2002[(data_europe2002['country']=='Spain') | (data_europe2002['country']=='France') | (data_europe2002['country']=='Germany') | (data_europe2002['country']=='Italy')]
name_gdp = data_fgis2002[['country', 'totalGDP']]
name_gdp.sort_values(by=['totalGDP'], ascending=False)


Spain experienced the greatest increase in total GDP between 2002 and 2007, increasing from .97 trillion USD to 1.16 trillion USD.

## 6

The '&' symbol is used to represent 'and' and would be used in cases where you are looking for data that has both of the given criteria. The '==' symbol is used for checking if a variable is equal to a certain value in conditional statements. Using a single equal sign would only be appropriate when assigning values to a variable name. The following code would return values where country is equal to Europe and the year is equal to 2007.

In [7]:
data_europe2007 = data[(data['continent']=='Europe') & (data['year'] == data['year'].max())]
data_europe2007

country continent  year  lifeExp       pop     gdpPercap
23                   Albania    Europe  2007   76.423   3600523   5937.029526
83                   Austria    Europe  2007   79.829   8199783  36126.492700
119                  Belgium    Europe  2007   79.441  10392226  33692.605080
155   Bosnia and Herzegovina    Europe  2007   74.852   4552198   7446.298803
191                 Bulgaria    Europe  2007   73.005   7322858  10680.792820
383                  Croatia    Europe  2007   75.748   4493312  14619.222720
407           Czech Republic    Europe  2007   76.486  10228744  22833.308510
419                  Denmark    Europe  2007   78.332   5468120  35278.418740
527                  Finland    Europe  2007   79.313   5238460  33207.084400
539                   France    Europe  2007   80.657  61083916  30470.016700
575                  Germany    Europe  2007   79.406  82400996  32170.374420
599                   Greece    Europe  2007   79.483  10706290  27538.411880
683                  Hungary    Europe  2007   73.338   9956108  18008.944440
695                  Iceland    Europe  2007   81.757    301931  36180.789190
755                  Ireland    Europe  2007   78.885   4109086  40675.996350
779                    Italy    Europe  2007   80.546  58147733  28569.719700
1019              Montenegro    Europe  2007   74.543    684736   9253.896111
1091             Netherlands    Europe  2007   79.762  16570613  36797.933320
1151                  Norway    Europe  2007   80.196   4627926  49357.190170
1235                  Poland    Europe  2007   75.563  38518241  15389.924680
1247                Portugal    Europe  2007   78.098  10642836  20509.647770
1283                 Romania    Europe  2007   72.476  22276056  10808.475610
1343                  Serbia    Europe  2007   74.002  10150265   9786.534714
1379         Slovak Republic    Europe  2007   74.663   5447502  18678.314350
1391                Slovenia    Europe  2007   77.926   2009245  25768.257590
1427                   Spain    Europe  2007   80.941  40448191  28821.063700
1475                  Sweden    Europe  2007   80.884   9031088  33859.748350
1487             Switzerland    Europe  2007   81.701   7554661  37506.419070
1583                  Turkey    Europe  2007   71.777  71158647   8458.276384
1607          United Kingdom    Europe  2007   79.425  60776238  33203.261280


The '|' symbol is used to represent 'or' and would return True as long as at least one of the given arguments is True. The following code will return True even though 3 is not greater than 4, because it is only asking if one of the given arguements is correct.

In [8]:
(1+2 == 3) | (3 > 4)

True

The '^' symbol is used to represent an exclusive or. It will return True if one argument is True and the other False, but will return False if both are True or both are False. The following will return False as both are False.

In [9]:
('cat' == 'dog') ^ (2 > 1)

True

## 7

The .loc command is a location command used to extract a row of data when given its index label as a parameter. The .iloc works similarly but returns a row by it's integer position in the dataset rather than its index label. The following code will return all rows of data between the integer positions 1691 and 1702, not including entry 1702.

In [11]:
data.iloc[1691:1702]

country continent  year  lifeExp       pop    gdpPercap
1691    Zambia    Africa  2007   42.384  11746035  1271.211593
1692  Zimbabwe    Africa  1952   48.451   3080907   406.884115
1693  Zimbabwe    Africa  1957   50.469   3646340   518.764268
1694  Zimbabwe    Africa  1962   52.358   4277736   527.272182
1695  Zimbabwe    Africa  1967   53.995   4995432   569.795071
1696  Zimbabwe    Africa  1972   55.635   5861135   799.362176
1697  Zimbabwe    Africa  1977   57.674   6642107   685.587682
1698  Zimbabwe    Africa  1982   60.363   7636524   788.855041
1699  Zimbabwe    Africa  1987   62.351   9216418   706.157306
1700  Zimbabwe    Africa  1992   60.377  10704340   693.420786
1701  Zimbabwe    Africa  1997   46.809  11404948   792.449960

The following code will return all observations from the second, third, and fourth columns. 

In [16]:
data.iloc[:, 1:4]

continent  year  lifeExp
0         Asia  1952   28.801
1         Asia  1957   30.332
2         Asia  1962   31.997
3         Asia  1967   34.020
4         Asia  1972   36.088
5         Asia  1977   38.438
6         Asia  1982   39.854
7         Asia  1987   40.822
8         Asia  1992   41.674
9         Asia  1997   41.763
10        Asia  2002   42.129
11        Asia  2007   43.828
12      Europe  1952   55.230
13      Europe  1957   59.280
14      Europe  1962   64.820
15      Europe  1967   66.220
16      Europe  1972   67.690
17      Europe  1977   68.930
18      Europe  1982   70.420
19      Europe  1987   72.000
20      Europe  1992   71.581
21      Europe  1997   72.950
22      Europe  2002   75.651
23      Europe  2007   76.423
24      Africa  1952   43.077
25      Africa  1957   45.685
26      Africa  1962   48.303
27      Africa  1967   51.407
28      Africa  1972   54.518
29      Africa  1977   58.014
...        ...   ...      ...
1674      Asia  1982   49.113
1675      Asia  1987   52.922
1676      Asia  1992   55.599
1677      Asia  1997   58.020
1678      Asia  2002   60.308
1679      Asia  2007   62.698
1680    Africa  1952   42.038
1681    Africa  1957   44.077
1682    Africa  1962   46.023
1683    Africa  1967   47.768
1684    Africa  1972   50.107
1685    Africa  1977   51.386
1686    Africa  1982   51.821
1687    Africa  1987   50.821
1688    Africa  1992   46.100
1689    Africa  1997   40.238
1690    Africa  2002   39.193
1691    Africa  2007   42.384
1692    Africa  1952   48.451
1693    Africa  1957   50.469
1694    Africa  1962   52.358
1695    Africa  1967   53.995
1696    Africa  1972   55.635
1697    Africa  1977   57.674
1698    Africa  1982   60.363
1699    Africa  1987   62.351
1700    Africa  1992   60.377
1701    Africa  1997   46.809
1702    Africa  2002   39.989
1703    Africa  2007   43.487

[1704 rows x 3 columns]

## 8

An ai is an Application Programming Interface, and it a part of a website's remote server that receives and processes requests. It allows different applications to work together.

To pull data from a remote server, you first have to send a request using the requests library.

In [10]:
import requests

You then must specify where the data is coming from in a url. 

In [11]:
url = "https://api.covidtracking.com/v1/states/daily.csv"

Then you can write it to a local file in a folder on your server using the os library.

In [12]:
import os

data_folder = 'data'
if not os.path.exists(data_folder):
    os.makedirs(data_folder)

After setting a filename for your data, you can retrieve the data from the request and populate your new file.

In [14]:
from datetime import datetime as dt
import pytz
 
file_name_short = 'ctp_' + str(dt.now(tz = pytz.utc)).replace(' ', '_') + '.csv'
file_name = os.path.join(data_folder, file_name_short)

r = requests.get(url)

with open(file_name, 'wb') as f:
    f.write(r.content)

Once your file is written, you cna import it into your workspace using pandas.

In [15]:
import pandas as pd
df = pd.read_csv(file_name)

## 9

The apply() function from the pandas library allows for the user to take any function and apply it to all values in the given series. For example, the apply() function could be used to sum all values for a particular column. Using apply() could be an alternative to writing out an actual function and then having to execute it across all series objects, which could lower output speed and efficiency, while taking up more space and increasing the chance of making a mistake. 

## 10 


Instead of using .iloc to filter columns, you could just make a new subset of your data frame. The following two lines of code return the same output; however, using the first option is potentially easier and allows both consecutive and non-consecutive columns to be extracted by name. 

In [16]:
data[["country", "continent", "year", "lifeExp"]]

country continent  year  lifeExp
0     Afghanistan      Asia  1952   28.801
1     Afghanistan      Asia  1957   30.332
2     Afghanistan      Asia  1962   31.997
3     Afghanistan      Asia  1967   34.020
4     Afghanistan      Asia  1972   36.088
5     Afghanistan      Asia  1977   38.438
6     Afghanistan      Asia  1982   39.854
7     Afghanistan      Asia  1987   40.822
8     Afghanistan      Asia  1992   41.674
9     Afghanistan      Asia  1997   41.763
10    Afghanistan      Asia  2002   42.129
11    Afghanistan      Asia  2007   43.828
12        Albania    Europe  1952   55.230
13        Albania    Europe  1957   59.280
14        Albania    Europe  1962   64.820
15        Albania    Europe  1967   66.220
16        Albania    Europe  1972   67.690
17        Albania    Europe  1977   68.930
18        Albania    Europe  1982   70.420
19        Albania    Europe  1987   72.000
20        Albania    Europe  1992   71.581
21        Albania    Europe  1997   72.950
22        Albania    Europe  2002   75.651
23        Albania    Europe  2007   76.423
24        Algeria    Africa  1952   43.077
25        Algeria    Africa  1957   45.685
26        Algeria    Africa  1962   48.303
27        Algeria    Africa  1967   51.407
28        Algeria    Africa  1972   54.518
29        Algeria    Africa  1977   58.014
...           ...       ...   ...      ...
1674  Yemen, Rep.      Asia  1982   49.113
1675  Yemen, Rep.      Asia  1987   52.922
1676  Yemen, Rep.      Asia  1992   55.599
1677  Yemen, Rep.      Asia  1997   58.020
1678  Yemen, Rep.      Asia  2002   60.308
1679  Yemen, Rep.      Asia  2007   62.698
1680       Zambia    Africa  1952   42.038
1681       Zambia    Africa  1957   44.077
1682       Zambia    Africa  1962   46.023
1683       Zambia    Africa  1967   47.768
1684       Zambia    Africa  1972   50.107
1685       Zambia    Africa  1977   51.386
1686       Zambia    Africa  1982   51.821
1687       Zambia    Africa  1987   50.821
1688       Zambia    Africa  1992   46.100
1689       Zambia    Africa  1997   40.238
1690       Zambia    Africa  2002   39.193
1691       Zambia    Africa  2007   42.384
1692     Zimbabwe    Africa  1952   48.451
1693     Zimbabwe    Africa  1957   50.469
1694     Zimbabwe    Africa  1962   52.358
1695     Zimbabwe    Africa  1967   53.995
1696     Zimbabwe    Africa  1972   55.635
1697     Zimbabwe    Africa  1977   57.674
1698     Zimbabwe    Africa  1982   60.363
1699     Zimbabwe    Africa  1987   62.351
1700     Zimbabwe    Africa  1992   60.377
1701     Zimbabwe    Africa  1997   46.809
1702     Zimbabwe    Africa  2002   39.989
1703     Zimbabwe    Africa  2007   43.487

[1704 rows x 4 columns]

In [17]:
data.iloc[:, 0:4]

country continent  year  lifeExp
0     Afghanistan      Asia  1952   28.801
1     Afghanistan      Asia  1957   30.332
2     Afghanistan      Asia  1962   31.997
3     Afghanistan      Asia  1967   34.020
4     Afghanistan      Asia  1972   36.088
5     Afghanistan      Asia  1977   38.438
6     Afghanistan      Asia  1982   39.854
7     Afghanistan      Asia  1987   40.822
8     Afghanistan      Asia  1992   41.674
9     Afghanistan      Asia  1997   41.763
10    Afghanistan      Asia  2002   42.129
11    Afghanistan      Asia  2007   43.828
12        Albania    Europe  1952   55.230
13        Albania    Europe  1957   59.280
14        Albania    Europe  1962   64.820
15        Albania    Europe  1967   66.220
16        Albania    Europe  1972   67.690
17        Albania    Europe  1977   68.930
18        Albania    Europe  1982   70.420
19        Albania    Europe  1987   72.000
20        Albania    Europe  1992   71.581
21        Albania    Europe  1997   72.950
22        Albania    Europe  2002   75.651
23        Albania    Europe  2007   76.423
24        Algeria    Africa  1952   43.077
25        Algeria    Africa  1957   45.685
26        Algeria    Africa  1962   48.303
27        Algeria    Africa  1967   51.407
28        Algeria    Africa  1972   54.518
29        Algeria    Africa  1977   58.014
...           ...       ...   ...      ...
1674  Yemen, Rep.      Asia  1982   49.113
1675  Yemen, Rep.      Asia  1987   52.922
1676  Yemen, Rep.      Asia  1992   55.599
1677  Yemen, Rep.      Asia  1997   58.020
1678  Yemen, Rep.      Asia  2002   60.308
1679  Yemen, Rep.      Asia  2007   62.698
1680       Zambia    Africa  1952   42.038
1681       Zambia    Africa  1957   44.077
1682       Zambia    Africa  1962   46.023
1683       Zambia    Africa  1967   47.768
1684       Zambia    Africa  1972   50.107
1685       Zambia    Africa  1977   51.386
1686       Zambia    Africa  1982   51.821
1687       Zambia    Africa  1987   50.821
1688       Zambia    Africa  1992   46.100
1689       Zambia    Africa  1997   40.238
1690       Zambia    Africa  2002   39.193
1691       Zambia    Africa  2007   42.384
1692     Zimbabwe    Africa  1952   48.451
1693     Zimbabwe    Africa  1957   50.469
1694     Zimbabwe    Africa  1962   52.358
1695     Zimbabwe    Africa  1967   53.995
1696     Zimbabwe    Africa  1972   55.635
1697     Zimbabwe    Africa  1977   57.674
1698     Zimbabwe    Africa  1982   60.363
1699     Zimbabwe    Africa  1987   62.351
1700     Zimbabwe    Africa  1992   60.377
1701     Zimbabwe    Africa  1997   46.809
1702     Zimbabwe    Africa  2002   39.989
1703     Zimbabwe    Africa  2007   43.487

[1704 rows x 4 columns]